packages/server/postgres/migrations/1724949259979_addFeatureFlagTables.ts

```diff
@@ -0,0 +1,73 @@
+import {Kysely, PostgresDialect, sql} from 'kysely'
+import getPg from '../getPg'
+
+export async function up() {
+  const pg = new Kysely<any>({
+    dialect: new PostgresDialect({
+      pool: getPg()
+    })
+  })
+
+  await pg.schema.createType('ScopeEnum').asEnum(['User', 'Team', 'Organization']).execute()
+
+  await pg.schema
+    .createTable('FeatureFlag')
+    .addColumn('id', 'uuid', (col) => col.primaryKey().defaultTo(sql`gen_random_uuid()`))
+    .addColumn('featureName', 'varchar(255)', (col) => col.notNull())
+    .addColumn('scope', sql`"ScopeEnum"`, (col) => col.notNull())
+    .addColumn('description', 'text')
+    .addColumn('expiresAt', 'timestamptz', (col) => col.notNull())
+    .addColumn('createdAt', 'timestamptz', (col) => col.notNull().defaultTo(sql`CURRENT_TIMESTAMP`))
+    .addColumn('updatedAt', 'timestamptz', (col) => col.notNull().defaultTo(sql`CURRENT_TIMESTAMP`))
+    .addUniqueConstraint('unique_featureName_scope', ['featureName', 'scope'])
+    .execute()
+
+  await pg.schema
+    .createTable('FeatureFlagOwner')
+    .addColumn('featureFlagId', 'uuid', (col) =>
+      col.notNull().references('FeatureFlag.id').onDelete('cascade')
+    )
+    .addColumn('userId', 'varchar(255)')
+    .addColumn('teamId', 'varchar(255)')
+    .addColumn('orgId', 'varchar(255)')
+    .addColumn('createdAt', 'timestamptz', (col) => col.notNull().defaultTo(sql`CURRENT_TIMESTAMP`))
+    .addCheckConstraint(
+      'check_feature_flag_owner_exclusivity',
+      sql`
+      (("userId" IS NOT NULL AND "teamId" IS NULL AND "orgId" IS NULL) OR
+       ("userId" IS NULL AND "teamId" IS NOT NULL AND "orgId" IS NULL) OR
+       ("userId" IS NULL AND "teamId" IS NULL AND "orgId" IS NOT NULL))
+    `
+    )
+    .execute()
+
+  await pg.schema
+    .createIndex('idx_feature_flag_owner_user')
+    .on('FeatureFlagOwner')
+    .columns(['userId', 'featureFlagId'])
+    .execute()
+
+  await pg.schema
+    .createIndex('idx_feature_flag_owner_team')
+    .on('FeatureFlagOwner')
+    .columns(['teamId', 'featureFlagId'])
+    .execute()
+
+  await pg.schema
+    .createIndex('idx_feature_flag_owner_org')
+    .on('FeatureFlagOwner')
+    .columns(['orgId', 'featureFlagId'])
+    .execute()
+}
```

Comment:
-1 I don't see how we would use these indexes.
I think what we want though is the same set of columns `['userId', 'featureFlagId']`, ... as unique constraints. Thanks to Matt I just learned that Postgres is treating `null` values as distinct by default[^1]

[^1]: https://www.postgresql.org/about/featurematrix/detail/392/